main.py

```diff
@@ -1,68 +1,225 @@
+Добро пожаловать в чат\\! Прочти правила, пожалуйста\\.
+このチャットへようこそ！ ルールをお読みください。
+
+[rules](https://t\\.me/dev\\_meme/3667)""",
+parse_mode=ParseMode.MARKDOWN_V2)
+
+def is_admin(chat: Chat, user: User) -> bool:
+	# might wanna cache admins
+	status = chat.get_member(user.id).status
+	return status == 'creator' or status == 'administrator'
+
+def get_warn_target(message: Message) -> User | None:
+	'''
+	Returns the user that is supposed to be warned. It might be a bot.
+	Returns None if no warn target.
+	'''
+	if message.reply_to_message is not None and message.reply_to_message.from_user is not None:
+		return message.reply_to_message.from_user
+	return None
+
+class WarnPossibleResult(NamedTuple):
+	'''
+	can_warn -> can warn happen (user is admin, user is not a bot, warn target isnt bot)
+	reason -> reason why warn cannot happen, might be None
+	'''
+	is_possible: bool
+	reason: str | None
+
+def is_warn_possible(message: Message, command: str) -> WarnPossibleResult:
+	if message.from_user.is_bot: # might be redundant
+		return WarnPossibleResult(False, None)
+	if not is_admin(message.chat, message.from_user):
+		return WarnPossibleResult(False, f'You are not an admin')
+	target = get_warn_target(message)
+	if target is None:
+		return WarnPossibleResult(False, f'Please reply to a message with /{command}')
+	if target.is_bot:
+		return WarnPossibleResult(False, f'Bots cannot be warned')
+	return WarnPossibleResult(True, None)
+
+@command("warn")
+@filter_chat(private_chat_id, private_chat_username)
+def warn_member(update: Update, context: CallbackContext):
+	warn_pos = is_warn_possible(update.message, 'warn')
+	if not warn_pos.is_possible:
+		if warn_pos.reason is not None:
+			update.message.reply_text(warn_pos.reason, parse_mode=ParseMode.MARKDOWN_V2)
+		return
+	target = get_warn_target(update.message)
+	userid = target.id
+	warns = db.get_warns(userid) + 1
+	db.set_warns(userid, warns)
+	update.message.chat.send_message(
+		f'*{get_mention(target)}* recieved a warn\\! Now they have {warns} warns',
+		parse_mode=ParseMode.MARKDOWN_V2)
+
+@command("unwarn")
+@filter_chat(private_chat_id, private_chat_username)
+def unwarn_member(update: Update, context: CallbackContext):
+	can_warn = is_warn_possible(update.message, 'unwarn')
+	if not can_warn.is_possible:
+		if can_warn.reason is not None:
+			update.message.reply_text(can_warn.reason, parse_mode=ParseMode.MARKDOWN_V2)
+		return
+	target = get_warn_target(update.message)
+	userid = target.id
+	warns = db.get_warns(userid)
+	if warns > 0:
+		warns -= 1
+	db.set_warns(userid, warns)
+	reply = f'*{target.mention_markdown_v2()}* been a good hooman\\! '
+	if warns == 0:
+		reply += 'Now they don\'t have any warns'
+	else:
+		reply += f'Now they have {warns} warns'
+	update.message.chat.send_message(reply,
+		parse_mode=ParseMode.MARKDOWN_V2)
+
+@command("clearwarns")
+@filter_chat(private_chat_id, private_chat_username)
+def unwarn_member(update: Update, context: CallbackContext):

```

Comment:
same function name as the function for /unwarn

maybe rename to clear_member_warns?

Reply:
Sure. Copy-paste error x)